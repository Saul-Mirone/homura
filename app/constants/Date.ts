export const format = 'LLL dd, yyyy';
