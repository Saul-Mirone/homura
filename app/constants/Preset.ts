export enum Preset {
  All = 'All Items',
  Archive = 'Archive',
  Starred = 'Starred',
  Unread = 'Unread',
}
