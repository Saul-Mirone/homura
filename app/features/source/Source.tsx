import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { channel } from '../../channel/child';
import { SideBar } from '../../components/SideBar';
import { Header } from '../../components/SideBar/Header';
import { SideBarItem } from '../../components/SideBar/SideBarItem';
import { AppDispatch } from '../../store';
import {
  asyncUpdateName,
  loadSource,
  selectSource,
  setActiveId,
} from './sourceSlice';

export const Source: React.FC<{ bottom: JSX.Element }> = ({ bottom }) => {
  const dispatch = useDispatch<AppDispatch>();
  const { list, activeId, mode, totalCount } = useSelector(selectSource);

  React.useEffect(() => {
    dispatch(loadSource(mode));
  }, [dispatch, mode]);

  const overview = (
    <Header
      mode={mode}
      active={activeId}
      count={totalCount}
      onClick={(id) => dispatch(setActiveId(id))}
    />
  );

  return (
    <SideBar
      onClick={() => {
        if (!activeId) return;
        dispatch(setActiveId());
      }}
      overview={overview}
      bottom={bottom}
    >
      {list.map(({ id, name, count, icon = undefined }) => (
        <SideBarItem
          onConfirmModify={(nextName) => {
            dispatch(asyncUpdateName(id, nextName));
          }}
          onUnsubscribe={() => {
            dispatch(setActiveId());
            channel.removeSourceById(id);
          }}
          key={id.toString()}
          url={icon}
          name={name}
          count={count}
          active={id === activeId}
          onClick={() => dispatch(setActiveId(id))}
        />
      ))}
    </SideBar>
  );
};
