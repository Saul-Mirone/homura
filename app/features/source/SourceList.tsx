import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { SideBar } from '../../components/SideBar';
import { Header } from '../../components/SideBar/Header';
import { SideBarItem } from '../../components/SideBar/SideBarItem';
import { AppDispatch } from '../../store';
import {
  fetchSources,
  selectSourceList,
  setCurrentSource,
  unsubscribeById,
  updateSourceById,
} from './sourceSlice';

export const SourceList: React.FC<{ bottom: JSX.Element }> = ({ bottom }) => {
  const dispatch = useDispatch<AppDispatch>();
  const { list, activeId, mode, totalCount } = useSelector(selectSourceList);

  React.useEffect(() => {
    dispatch(fetchSources(mode));
  }, [dispatch, mode]);

  const overview = (
    <Header
      mode={mode}
      active={activeId}
      count={totalCount}
      onClick={(id) => dispatch(setCurrentSource(id))}
    />
  );

  return (
    <SideBar
      onClick={() => dispatch(setCurrentSource(null))}
      overview={overview}
      bottom={bottom}
    >
      {list.map(({ id, name, count, icon }) => (
        <SideBarItem
          key={id.toString()}
          id={id}
          url={icon}
          name={name}
          count={count}
          active={id === activeId}
          onConfirmModify={(nextName) =>
            dispatch(updateSourceById({ id, name: nextName }))
          }
          onUnsubscribe={() => dispatch(unsubscribeById(id))}
          onClick={() => dispatch(setCurrentSource(id))}
        />
      ))}
    </SideBar>
  );
};
