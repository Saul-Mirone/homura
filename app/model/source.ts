import {
  Association,
  DataTypes,
  HasManyAddAssociationMixin,
  HasManyCountAssociationsMixin,
  HasManyCreateAssociationMixin,
  HasManyGetAssociationsMixin,
  HasManyHasAssociationMixin,
  Model,
  Optional,
  Sequelize,
} from 'sequelize';
import { Post } from './post';

export interface SourceAttributes {
  id: number;
  name: string;
  link: string;
  sourceUrl: string;
  icon: string | null;
}

export type CreateSourceAttributes = Optional<SourceAttributes, 'id'>;

export class Source
  extends Model<SourceAttributes, CreateSourceAttributes>
  implements SourceAttributes {
  public id!: number;

  public name!: string;

  public link!: string;

  public sourceUrl!: string;

  public icon!: string | null;

  public getPosts!: HasManyGetAssociationsMixin<Post>;

  public addPost!: HasManyAddAssociationMixin<Post, number>;

  public hasPost!: HasManyHasAssociationMixin<Post, number>;

  public countPosts!: HasManyCountAssociationsMixin;

  public createPost!: HasManyCreateAssociationMixin<Post>;

  public posts?: Post[];

  public toJSON!: () => SourceAttributes & {
    posts: Array<ReturnType<Post['toJSON']>>;
  };

  public static associations: {
    posts: Association<Source, Post>;
  };
}

export const initSource = (sequelize: Sequelize): void => {
  Source.init(
    {
      id: {
        type: DataTypes.INTEGER,
        autoIncrement: true,
        primaryKey: true,
      },
      sourceUrl: {
        type: DataTypes.STRING,
        allowNull: false,
      },
      name: {
        type: DataTypes.STRING,
        allowNull: false,
      },
      icon: {
        type: DataTypes.STRING,
        allowNull: true,
      },
      link: {
        type: DataTypes.STRING,
        allowNull: false,
      },
    },
    { sequelize, tableName: 'sources' }
  );

  Source.hasMany(Post, {
    sourceKey: 'id',
    foreignKey: 'sourceId',
    as: 'posts',
  });
  Post.belongsTo(Source, {
    targetKey: 'id',
    foreignKey: 'sourceId',
    as: 'source',
  });
};
