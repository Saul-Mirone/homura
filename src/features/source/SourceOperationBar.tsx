import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { BottomBar, Step } from '../../components/SideBar/BottomBar';
import { FeedSearchBar } from '../../components/SideBar/FeedSearchBar';
import { FeedSubscribeBar } from '../../components/SideBar/FeedSubscribeBar';
import { AppDispatch } from '../../store';
import {
  resetSubscribeError,
  resetSubscribeState,
  searchUrlForSource,
  selectSourceOperation,
  showSubscribeBar,
  subscribeToSource,
  syncSources,
} from './sourceSlice';

export const SourceOperationBar: React.FC = () => {
  const dispatch = useDispatch<AppDispatch>();
  const {
    mode,
    loading,
    subscribeStep,
    subscribeName,
    subscribeLink,
    subscribeError,
  } = useSelector(selectSourceOperation);

  const renderBottom = React.useCallback(
    (currentStep: Step) => {
      switch (currentStep) {
        case Step.EnterUrl: {
          return (
            <FeedSearchBar
              hasError={subscribeError === 'ParseRSSFailed'}
              loading={loading}
              onSearch={(inputLink) => dispatch(searchUrlForSource(inputLink))}
              onCancel={() => dispatch(resetSubscribeState())}
              onClickError={() => dispatch(resetSubscribeError())}
            />
          );
        }
        case Step.EnterName: {
          return (
            <FeedSubscribeBar
              link={subscribeLink}
              initialName={subscribeName}
              onCancel={() => dispatch(resetSubscribeState())}
              onConfirm={(inputName) => {
                dispatch(
                  subscribeToSource({
                    name: inputName,
                    mode,
                  })
                );
              }}
            />
          );
        }
        default: {
          throw new Error('Invalid Step');
        }
      }
    },
    [subscribeError, subscribeLink, subscribeName, dispatch, loading, mode]
  );

  return (
    <BottomBar
      step={subscribeStep}
      loading={loading}
      onClickPlus={() => dispatch(showSubscribeBar())}
      onClickSync={() => dispatch(syncSources())}
      render={renderBottom}
    />
  );
};
