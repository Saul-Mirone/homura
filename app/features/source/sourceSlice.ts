import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { channel } from '../../channel/child';
import { Step } from '../../components/SideBar/BottomBar';
import { Mode } from '../../constants/Mode';
import { Preset } from '../../constants/Preset';
import { Status } from '../../constants/Status';
import type { RootState } from '../../store';

type SourceItem = {
  id: number;
  name: string;
  link: string;
  count: number;
  icon: string | null;
};

type SourceListState = {
  activeId: number | Preset | null;
  list: SourceItem[];
  fetchListStatus: Status;
};

type SourceSubscribeState = {
  subscribeLink: string;
  subscribeName: string;
  subscribeStep: Step | null;
  subscribeStatus: Status;
  subscribeError: null | string;
};

export type State = SourceListState & SourceSubscribeState;

const initialSourceList: SourceListState = {
  list: [],
  activeId: null,
  fetchListStatus: Status.Idle,
};

const initialSourceSubscribeState: SourceSubscribeState = {
  subscribeLink: '',
  subscribeName: '',
  subscribeStep: null,
  subscribeStatus: Status.Idle,
  subscribeError: null,
};

const initialState: State = {
  ...initialSourceList,
  ...initialSourceSubscribeState,
};

const fetchSourceListAPI = async (mode: Mode) => {
  const countType = mode === Mode.Starred ? 'starred' : 'unread';
  const data = await channel.getSourceList(countType);
  return data.map(({ id, name, icon, link, count }) => ({
    id,
    name,
    icon,
    link,
    count,
  }));
};

export const fetchSources = createAsyncThunk(
  'source/fetchSources',
  (mode: Mode) => {
    return fetchSourceListAPI(mode);
  }
);

export const syncSources = createAsyncThunk(
  'source/syncSources',
  async (mode: Mode) => {
    await channel.sync();
    return fetchSourceListAPI(mode);
  }
);

export const unsubscribeById = createAsyncThunk(
  'source/unsubscribeById',
  async (id: number) => {
    await channel.removeSourceById(id);
    return id;
  }
);

export const updateSourceById = createAsyncThunk(
  'source/updateSourceById',
  async ({ id, name }: { id: number; name: string }) => {
    await channel.updateSourceNameById(id, name);

    return { id, name };
  }
);

export const searchUrlForSource = createAsyncThunk(
  'source/searchUrlForSource',
  async (link: string) => {
    const result = await channel.checkUrl(link);
    if (!result) {
      throw new Error('ParseRSSFailed');
    }
    return result;
  }
);

export const subscribeToSource = createAsyncThunk(
  'source/subscribeToSource',
  async ({ name, mode }: { name: string; mode: Mode }) => {
    const { posts, icon, link, id } = await channel.confirm(name);

    return {
      id,
      name,
      link,
      icon,
      count: mode === Mode.Starred ? 0 : posts.length,
    };
  }
);

const sourceSlice = createSlice({
  name: 'source',
  initialState,
  reducers: {
    setCurrentSource: (
      state,
      action: PayloadAction<number | Preset | null>
    ) => {
      state.activeId = action.payload;
    },
    showSubscribeBar: (state) => {
      state.subscribeStep = Step.EnterUrl;
      state.subscribeStatus = Status.Idle;
    },
    resetSubscribeError: (state) => {
      state.subscribeError = null;
    },
    resetSubscribeState: (state) => {
      state.subscribeStep = null;
      state.subscribeError = null;
      state.subscribeLink = '';
      state.subscribeName = '';
      state.subscribeStatus = Status.Idle;
    },

    // TODO: need to be reviewed
    countDownOne: (state, action: PayloadAction<number>) => {
      const target = state.list.find((x) => x.id === action.payload);
      if (!target) return;
      target.count -= 1;
    },

    // TODO: need to be reviewed
    countUpOne: (state, action: PayloadAction<number>) => {
      const target = state.list.find((x) => x.id === action.payload);
      if (!target) return;
      target.count += 1;
    },

    // TODO: need to be reviewed
    countToZero: (state, action: PayloadAction<number>) => {
      const target = state.list.find((x) => x.id === action.payload);
      if (!target) return;
      target.count = 0;
    },
  },
  extraReducers: (builder) =>
    builder
      .addCase(fetchSources.pending, (state) => {
        state.fetchListStatus = Status.Pending;
      })
      .addCase(fetchSources.fulfilled, (state, action) => {
        state.fetchListStatus = Status.Succeeded;
        state.list = action.payload;
      })
      .addCase(fetchSources.rejected, (state) => {
        state.fetchListStatus = Status.Failed;
      })
      .addCase(syncSources.pending, (state) => {
        state.fetchListStatus = Status.Pending;
      })
      .addCase(syncSources.fulfilled, (state, action) => {
        state.fetchListStatus = Status.Succeeded;
        state.list = action.payload;
      })
      .addCase(syncSources.rejected, (state) => {
        state.fetchListStatus = Status.Failed;
      })
      .addCase(unsubscribeById.fulfilled, (state, action) => {
        state.activeId = null;
        state.list = state.list.filter((x) => x.id !== action.payload);
      })
      .addCase(updateSourceById.fulfilled, (state, action) => {
        state.activeId = null;
        const target = state.list.find((x) => x.id === action.payload.id);
        if (!target) return;
        target.name = action.payload.name;
      })
      .addCase(searchUrlForSource.pending, (state, action) => {
        state.subscribeLink = action.meta.arg;
        state.subscribeStatus = Status.Pending;
      })
      .addCase(searchUrlForSource.fulfilled, (state, action) => {
        state.subscribeName = action.payload;
        state.subscribeStep = Step.EnterName;
      })
      .addCase(searchUrlForSource.rejected, (state, action) => {
        state.subscribeStatus = Status.Failed;
        state.subscribeError = action.error.message ?? 'Unknown Error';
      })
      .addCase(subscribeToSource.fulfilled, (state, action) => {
        const { id } = action.payload;
        state.subscribeStatus = Status.Succeeded;
        state.list.push(action.payload);
        state.activeId = id;
        state.subscribeStep = null;
        state.subscribeError = null;
        state.subscribeLink = '';
        state.subscribeName = '';
      }),
});

export const {
  setCurrentSource,
  resetSubscribeError,
  resetSubscribeState,
  showSubscribeBar,

  countDownOne,
  countUpOne,
  countToZero,
} = sourceSlice.actions;

export const sourceReducer = sourceSlice.reducer;

export const selectSourceList = (state: RootState) => {
  const { source, mode } = state;
  const totalCount = source.list.reduce((acc, cur) => acc + cur.count, 0);
  const { list, activeId } = source;

  return {
    activeId,
    totalCount,
    mode,
    list: list.filter((x) => {
      if (mode === Mode.All || activeId === x.id) return true;
      return x.count > 0;
    }),
  };
};

export const selectSourceOperation = (state: RootState) => {
  const { source, mode } = state;
  const {
    subscribeLink,
    subscribeName,
    subscribeStep,
    subscribeStatus,
    subscribeError,
    fetchListStatus,
  } = source;
  const loading =
    subscribeStatus === Status.Pending || fetchListStatus === Status.Pending;
  return {
    mode,
    subscribeLink,
    subscribeName,
    subscribeStep,
    subscribeError,
    loading,
  };
};
