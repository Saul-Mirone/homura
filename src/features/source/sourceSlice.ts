import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { channel } from '../../channel/child';
import { Step } from '../../components/SideBar/BottomBar';
import { Mode } from '../../constants/Mode';
import { Preset } from '../../constants/Preset';
import { Status } from '../../constants/Status';
import { Source } from '../../model/types';
import type { RootState } from '../../store';

type SourceKeys = 'id' | 'name' | 'link' | 'icon';
type SourceItem = Pick<Source, SourceKeys> & { count: number };

type SourceListState = {
  activeId: number | Preset | undefined;
  list: SourceItem[];
  fetchListStatus: Status;
  syncListStatus: Status;
};

type SourceSubscribeState = {
  subscribeLink: string;
  subscribeName: string;
  subscribeStatus: Status;
  subscribeStep: Step | undefined;
  subscribeError: string | undefined;
};

export type State = SourceListState & SourceSubscribeState;

const initialSourceList: SourceListState = {
  list: [],
  activeId: undefined,
  fetchListStatus: Status.Idle,
  syncListStatus: Status.Idle,
};

const initialSourceSubscribeState: SourceSubscribeState = {
  subscribeLink: '',
  subscribeName: '',
  subscribeStep: undefined,
  subscribeStatus: Status.Idle,
  subscribeError: undefined,
};

const initialState: State = {
  ...initialSourceList,
  ...initialSourceSubscribeState,
};

const fetchSourceListAPI = (mode: Mode) => {
  const countType = mode === Mode.Starred ? 'starred' : 'unread';
  return channel.getSourceList(countType);
};

export const fetchSources = createAsyncThunk(
  'source/fetchSources',
  (mode: Mode) => fetchSourceListAPI(mode)
);

export const syncSources = createAsyncThunk('source/syncSources', () =>
  channel.sync()
);

export const unsubscribeById = createAsyncThunk(
  'source/unsubscribeById',
  async (id: number) => {
    await channel.removeSourceById(id);
    return id;
  }
);

export const updateSourceById = createAsyncThunk(
  'source/updateSourceById',
  async ({ id, name }: { id: number; name: string }) => {
    await channel.updateSourceNameById(id, name);

    return { id, name };
  }
);

export const searchUrlForSource = createAsyncThunk(
  'source/searchUrlForSource',
  async (link: string) => {
    const result = await channel.checkUrl(link);
    if (!result) {
      throw new Error('ParseRSSFailed');
    }
    return result;
  }
);

export const subscribeToSource = createAsyncThunk(
  'source/subscribeToSource',
  async ({ name, mode }: { name: string; mode: Mode }) => {
    const { count, icon, link, id } = await channel.confirm(name);

    return {
      id,
      name,
      link,
      icon,
      count: mode === Mode.Starred ? 0 : count,
    };
  }
);

const sourceSlice = createSlice({
  name: 'source',
  initialState,
  reducers: {
    setCurrentSource: (
      state,
      action: PayloadAction<number | Preset | undefined>
    ) => {
      state.activeId = action.payload;
    },
    showSubscribeBar: (state) => {
      state.subscribeStep = Step.EnterUrl;
      state.subscribeStatus = Status.Idle;
    },
    resetSubscribeError: (state) => {
      state.subscribeError = undefined;
    },
    resetSubscribeState: (state) => {
      state.subscribeStep = undefined;
      state.subscribeError = undefined;
      state.subscribeLink = '';
      state.subscribeName = '';
      state.subscribeStatus = Status.Idle;
    },

    setCountToZero: (state, action: PayloadAction<number>) => {
      const target = state.list.find((x) => x.id === action.payload);
      if (!target) return;
      target.count = 0;
    },
    incCount: (state, action: PayloadAction<number>) => {
      const target = state.list.find((x) => x.id === action.payload);
      if (!target) return;
      target.count += 1;
    },
    decCount: (state, action: PayloadAction<number>) => {
      const target = state.list.find((x) => x.id === action.payload);
      if (!target) return;
      target.count -= 1;
    },
  },
  extraReducers: (builder) =>
    builder
      .addCase(fetchSources.pending, (state) => {
        state.fetchListStatus = Status.Pending;
      })
      .addCase(fetchSources.fulfilled, (state, action) => {
        state.fetchListStatus = Status.Succeeded;
        state.list = action.payload;
      })
      .addCase(fetchSources.rejected, (state) => {
        state.fetchListStatus = Status.Failed;
      })
      .addCase(syncSources.pending, (state) => {
        state.syncListStatus = Status.Pending;
      })
      .addCase(syncSources.fulfilled, (state) => {
        state.syncListStatus = Status.Succeeded;
      })
      .addCase(syncSources.rejected, (state) => {
        state.syncListStatus = Status.Failed;
      })
      .addCase(unsubscribeById.fulfilled, (state, action) => {
        state.activeId = undefined;
        state.list = state.list.filter((x) => x.id !== action.payload);
      })
      .addCase(updateSourceById.fulfilled, (state, action) => {
        state.activeId = undefined;
        const target = state.list.find((x) => x.id === action.payload.id);
        if (!target) return;
        target.name = action.payload.name;
      })
      .addCase(searchUrlForSource.pending, (state, action) => {
        state.subscribeLink = action.meta.arg;
        state.subscribeStatus = Status.Pending;
      })
      .addCase(searchUrlForSource.fulfilled, (state, action) => {
        state.subscribeName = action.payload;
        state.subscribeStep = Step.EnterName;
      })
      .addCase(searchUrlForSource.rejected, (state, action) => {
        state.subscribeStatus = Status.Failed;
        state.subscribeError = action.error.message ?? 'Unknown Error';
      })
      .addCase(subscribeToSource.fulfilled, (state, action) => {
        const { id } = action.payload;
        state.subscribeStatus = Status.Succeeded;
        state.list.push(action.payload);
        state.activeId = id;
        state.subscribeStep = undefined;
        state.subscribeError = undefined;
        state.subscribeLink = '';
        state.subscribeName = '';
      }),
});

export const {
  setCurrentSource,
  resetSubscribeError,
  resetSubscribeState,
  showSubscribeBar,
  setCountToZero,
  decCount,
  incCount,
} = sourceSlice.actions;

export const sourceReducer = sourceSlice.reducer;

export const selectSourceList = (state: RootState) => {
  const { source, mode } = state;
  const totalCount = source.list.reduce((acc, cur) => acc + cur.count, 0);
  const { list, activeId, syncListStatus } = source;

  return {
    activeId,
    totalCount,
    mode,
    syncStatus: syncListStatus,
    list: list.filter((x) => {
      if (mode === Mode.All || activeId === x.id) return true;
      return x.count > 0;
    }),
  };
};

export const selectSourceOperation = (state: RootState) => {
  const { source, mode } = state;
  const {
    subscribeLink,
    subscribeName,
    subscribeStep,
    subscribeStatus,
    subscribeError,
    fetchListStatus,
    syncListStatus,
  } = source;
  const loading =
    subscribeStatus === Status.Pending ||
    fetchListStatus === Status.Pending ||
    syncListStatus === Status.Pending;
  return {
    mode,
    subscribeLink,
    subscribeName,
    subscribeStep,
    subscribeError,
    loading,
  };
};
