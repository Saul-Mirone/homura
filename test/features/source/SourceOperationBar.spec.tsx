import { configureStore } from '@reduxjs/toolkit';
import { render, screen } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import { Status } from '../../../src/constants/Status';
import { SourceOperationBar } from '../../../src/features/source/SourceOperationBar';
import * as sourceSlice from '../../../src/features/source/sourceSlice';

function setup(
  preloadedState: {
    source: Partial<sourceSlice.State>;
  } = {
    source: {},
  }
) {
  const initialSourceState = {
    list: [],
    activeId: undefined,
    fetchListStatus: Status.Idle,
    syncListStatus: Status.Idle,
    subscribeLink: '',
    subscribeName: '',
    subscribeStep: undefined,
    subscribeStatus: Status.Idle,
    subscribeError: undefined,
  };
  const state: {
    source: sourceSlice.State;
  } = {
    source: { ...initialSourceState, ...preloadedState.source },
  };
  const store = configureStore({
    reducer: { source: sourceSlice.sourceReducer },
    preloadedState: state,
  });

  render(
    <Provider store={store}>
      <SourceOperationBar />
    </Provider>
  );

  return {
    store,
  };
}

describe('SourceOperationBar component', () => {
  it('should match snapshot when list is empty', () => {
    setup();

    expect(screen.getByRole('toolbar')).toMatchSnapshot();
  });
});
