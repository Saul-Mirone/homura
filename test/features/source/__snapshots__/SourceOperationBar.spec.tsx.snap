// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SourceOperationBar component should match snapshot when list is empty 1`] = `
<div
  class="sidebar-bottom-bar"
  role="toolbar"
>
  <div
    class="p-2  "
    role="button"
    tabindex="0"
  >
    <div
      class="w-5 h-5 m-auto  "
      role="button"
      tabindex="0"
    >
      <svg
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 6v6m0 0v6m0-6h6m-6 0H6"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
  </div>
  <div
    class="p-2  "
    role="button"
    tabindex="0"
  >
    <div
      class="w-5 h-5 m-auto  "
      role="button"
      tabindex="0"
    >
      <svg
        class=""
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4 4v5h.582m15.356 2A8.001 8.001 0 004.582 9m0 0H9m11 11v-5h-.581m0 0a8.003 8.003 0 01-15.357-2m15.357 2H15"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
  </div>
</div>
`;
